resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopGlobalFeatureDriverStep.scala

```diff
@@ -0,0 +1,151 @@
+        " you must also specify the name of the secret",
+      "If a secret storing a Kerberos Delegation Token is specified you must also" +
+        " specify the item-key where the data is stored")
+
+    require(kubernetesConf.hadoopConfDir.isDefined, "Ensure that HADOOP_CONF_DIR is defined")
+    private val hadoopConfDir = kubernetesConf.hadoopConfDir.get
+    private val hadoopConfigurationFiles = kubeTokenManager.getHadoopConfFiles(hadoopConfDir)
+
+    // Either use pre-existing secret or login to create new Secret with DT stored within
+    private val hadoopSpec: Option[KerberosConfigSpec] = (for {
+      secretName <- maybeExistingSecretName
+      secretItemKey <- maybeExistingSecretItemKey
+    } yield {
+      KerberosConfigSpec(
+         dtSecret = None,
+         dtSecretName = secretName,
+         dtSecretItemKey = secretItemKey,
+         jobUserName = kubeTokenManager.getCurrentUser.getShortUserName)
+    }).orElse(
+      if (isKerberosEnabled) {
+         Some(HadoopKerberosLogin.buildSpec(
+             conf,
+             kubernetesConf.appResourceNamePrefix,
+             kubeTokenManager))
+       } else None )
+
+    override def configurePod(pod: SparkPod): SparkPod = {
+      val hadoopBasedSparkPod = HadoopBootstrapUtil.bootstrapHadoopConfDir(
+        hadoopConfDir,
+        kubernetesConf.hadoopConfigMapName,
+        kubeTokenManager,
+        pod)
+      (for {
+        hSpec <- hadoopSpec
+        krb5fileLocation <- kubernetesConf.get(KUBERNETES_KERBEROS_KRB5_FILE)
+      } yield {
+        HadoopBootstrapUtil.bootstrapKerberosPod(
+            hSpec.dtSecretName,
+            hSpec.dtSecretItemKey,
+            hSpec.jobUserName,
+            krb5fileLocation,
+            kubernetesConf.kRBConfigMapName,
+            hadoopBasedSparkPod)
+      }).getOrElse(
+        HadoopBootstrapUtil.bootstrapSparkUserPod(
+          kubeTokenManager.getCurrentUser.getShortUserName,
+          hadoopBasedSparkPod))
+    }
+
+    override def getAdditionalPodSystemProperties(): Map[String, String] = {
+      val resolvedConfValues = hadoopSpec.map{ hSpec =>
+         Map(KERBEROS_KEYTAB_SECRET_NAME -> hSpec.dtSecretName,
+            KERBEROS_KEYTAB_SECRET_KEY -> hSpec.dtSecretItemKey,
+            KERBEROS_SPARK_USER_NAME -> hSpec.jobUserName)
+      }.getOrElse(
+          Map(KERBEROS_SPARK_USER_NAME ->
+            kubernetesConf.tokenManager.getCurrentUser.getShortUserName))
+      Map(HADOOP_CONFIG_MAP_SPARK_CONF_NAME -> kubernetesConf.hadoopConfigMapName,
+          HADOOP_CONF_DIR_LOC -> kubernetesConf.hadoopConfDir.get) ++ resolvedConfValues
+    }
+
+    override def getAdditionalKubernetesResources(): Seq[HasMetadata] = {
+      val krb5ConfigMap = kubernetesConf.get(KUBERNETES_KERBEROS_KRB5_FILE)
+        .map(fileLocation => HadoopBootstrapUtil.buildkrb5ConfigMap(
+          kubernetesConf.kRBConfigMapName,
+          fileLocation))
+      val kerberosDTSecret = for {
```

Comment:
Specifically here you can do `hSpec.dtSecret.filter( _ => hadoopSpec.isDefined)`

Reply:
In fact here you can just do `val dtSecret = hadoopSpec.map(_.dtSecret)`

Most places where I've seen `for...yield` here can be replaced with `option.map { ... }`.

Reply:
The above recommendations give a `Option[Option[Secret]]`. I was aware of this, but the below resolution of `kerberosDTSecret.toSeq` would not be resolved correctly with the above recommendations, as it is looking for `Option[Secret]`. As such, I think the `for...yield` is the correct way of approaching this and is idiomatic in Scala.